### PR TITLE
add balancer v2 pools for gnosis and other supported evm chains.

### LINKED ISSUE #7405: Support balancer pools in gnosis chain and more
## Abstract

Support balancer pool decoding in gnosis chain and all other chains we don't support yet. A related issue is to automatically detect LP positions and their value: https://github.com/rotki/rotki/issues/5208

## Motivation

Gnosis chain is getting a lot of traction right now and we should support more protocols there.

### PR DESCRIPTION
**Pull request by prettyirrelevant**
Closes #7405 

this branch is based on #8894. it introduces balancer v2 decoding & lp balances for supported evm chains.

### CODE REVIEW

File: rotkehlchen/chain/evm/decoding/balancer/mixins.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+from rotkehlchen.chain.evm.structures import EvmTxReceiptLog
+from rotkehlchen.history.events.structures.types import HistoryEventSubType, HistoryEventType
+from rotkehlchen.types import EvmTransaction
+
+if TYPE_CHECKING:
+    from rotkehlchen.history.events.structures.evm_event import EvmEvent
+
+
+class BalancerCommonAccountingMixin:
+    def __init__(self, counterparty: str) -> None:
```
Comment (LefterisJP) [inline rotkehlchen/chain/evm/decoding/balancer/mixins.py:14]: Write out the possible types

```suggestion
    def __init__(self, counterparty: Literal['choice1', 'choice2']) -> None:
```

File: rotkehlchen/chain/evm/decoding/balancer/mixins.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+        self.counterparty = counterparty
+
+    def _check_deposits_withdrawals(
+            self,
+            transaction: EvmTransaction,  # pylint: disable=unused-argument
+            decoded_events: list['EvmEvent'],
+            all_logs: list[EvmTxReceiptLog],  # pylint: disable=unused-argument
+    ) -> list['EvmEvent']:
+        """Check for accounting in balancer that the deposits/withdrawals events have the needed information to process them during accounting."""  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/chain/evm/decoding/balancer/mixins.py:23]: Hmm this sentence does not make sense. What do you want to write?

File: rotkehlchen/chain/evm/decoding/balancer/mixins.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+            # When joining a pool:
+            # - V1: first we spend the assets, then we receive the BPT token
+            # - V2: first we receive the BPT token, then we spend the assets
+            # When exiting a pool:
+            # First we return the BPT token, then we receive the assets (both V1 and V2)
+            if (
+                event.event_type == HistoryEventType.RECEIVE and
+                event.event_subtype == HistoryEventSubType.RECEIVE_WRAPPED
+            ):
+                # For V1, we map the deposit events to the pool token event
```
Comment (LefterisJP) [inline rotkehlchen/chain/evm/decoding/balancer/mixins.py:42]:
```suggestion
            ):  # For V1, we map the deposit events to the pool token event
```

File: rotkehlchen/chain/evm/decoding/balancer/mixins.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+                last_event = event
+
+            elif ((
+                event.event_type == HistoryEventType.DEPOSIT and
+                event.event_subtype == HistoryEventSubType.DEPOSIT_ASSET
+            ) or (
+                event.event_type == HistoryEventType.WITHDRAWAL and
+                event.event_subtype == HistoryEventSubType.REFUND
+            )):
+                # Handle direction change from spending to receiving assets
```
Comment (LefterisJP) [inline rotkehlchen/chain/evm/decoding/balancer/mixins.py:66]:
```suggestion
            )):  # Handle direction change from spending to receiving assets
```

File: rotkehlchen/chain/evm/decoding/balancer/mixins.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+                if last_event is not None and last_event.event_type == HistoryEventType.SPEND:
+                    # save the exit event and reset the related events
+                    related_events_map[last_event] = related_events
+                    last_event = None
+                    related_events = []
+
+                related_events.append(event)
+
+                # For V2 joins, map deposit events to the preceding pool token event
+                if (
```
Comment (LefterisJP) [inline rotkehlchen/chain/evm/decoding/balancer/mixins.py:76]:
```suggestion
                
                if (  # For V2 joins, map deposit events to the preceding pool token event
```

File: rotkehlchen/chain/evm/decoding/balancer/utils.py [hunk 6]
```diff
@@ -41,17 +42,18 @@
 
 def query_balancer_v1_data(
         inquirer: 'EvmNodeInquirer',
-        cache_type: Literal[CacheType.BALANCER_V1_POOLS],
+        cache_type: Literal[CacheType.BALANCER_V1_POOLS, CacheType.BALANCER_V2_POOLS],
```
Comment (LefterisJP) [inline rotkehlchen/chain/evm/decoding/balancer/utils.py:45]: Pass `version` and `protocol in this function. Also rename it. It's no longer just for v1, right?

This way you will not need to do these if checks you are doing below in the start and in each iteration of the loop

File: rotkehlchen/chain/evm/decoding/balancer/utils.py [hunk 7]
```diff
@@ -41,17 +42,18 @@
...
 ) -> None:
     """Query and store balancer v1 pools.
     May raise:
     - RemoteError
     """
+    version = 1 if cache_type == CacheType.BALANCER_V1_POOLS else 2
```
Comment (LefterisJP) [inline rotkehlchen/chain/evm/decoding/balancer/utils.py:51]: this should go away

File: rotkehlchen/chain/evm/decoding/balancer/utils.py [hunk 8]
```diff
@@ -72,30 +74,32 @@ def query_balancer_v1_data(
             )
             return None
 
-    for pool in _query_balancer_pools(chain=inquirer.chain_id, version=1):
+    for pool in _query_balancer_pools(chain=inquirer.chain_id, version=version):
+        protocol = CPT_BALANCER_V1 if version == 1 else CPT_BALANCER_V2
```
Comment (LefterisJP) [inline rotkehlchen/chain/evm/decoding/balancer/utils.py:78]: this should go away

File: rotkehlchen/chain/evm/decoding/balancer/utils.py [hunk 9]
```diff
@@ -155,7 +159,7 @@ def query_balancer_api(query: str, variables: dict[str, Any]) -> dict[str, Any]:
         raise RemoteError(f'Balancer API request with query={query} and vars={variables} failed due to {msg}') from e  # noqa: E501
 
 
-def _query_balancer_pools_count(chain: 'ChainID', version: Literal[1, 2]) -> int:
+def _query_balancer_pools_count(chain: 'ChainID', version: int) -> int:
```
Comment (LefterisJP) [inline rotkehlchen/chain/evm/decoding/balancer/utils.py:162]: why change the type? Isn't it either 1 or 2?

File: rotkehlchen/chain/evm/decoding/balancer/v2/balances.py [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+            tx_decoder: 'EVMTransactionDecoder',
+    ):
+        super().__init__(
+            evm_inquirer=evm_inquirer,
+            tx_decoder=tx_decoder,
+            counterparty=CPT_BALANCER_V2,
+            deposit_event_types={(HistoryEventType.DEPOSIT, HistoryEventSubType.DEPOSIT_ASSET)},
+        )
+
+    def query_balances(self) -> 'BalancesSheetType':
```
Comment (LefterisJP) [inline rotkehlchen/chain/evm/decoding/balancer/v2/balances.py:45]: isn't this function very similart ot v1 balances? Can't you just simply abstract it somehow?

File: rotkehlchen/db/upgrades/v45_v46.py [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+
+        try:
+            active_modules = json.loads(active_modules_result[0])
+        except json.JSONDecodeError:
+            log.error(f'During v45->v46 DB upgrade a non-json active_modules entry was found: {active_modules_result[0]}.')  # noqa: E501
+            return None
+
+        write_cursor.execute(
+            "UPDATE OR IGNORE settings SET value=? WHERE name='active_modules'",
+            (json.dumps([module for module in active_modules if module != 'balancer']),),
```
Comment (LefterisJP) [inline rotkehlchen/db/upgrades/v45_v46.py:36]: is it only balancer what we care about? There is no v1 or v2 there right?

File: rotkehlchen/tests/db/test_db_upgrades.py [hunk 12]
```diff
@@ -2596,6 +2596,36 @@ def test_upgrade_db_44_to_45(user_data_dir, messages_aggregator):
...
+    _use_prepared_db(user_data_dir, 'v45_rotkehlchen.db')
+    db_v5 = _init_db_with_target_version(
+        target_version=45,
+        user_data_dir=user_data_dir,
+        msg_aggregator=messages_aggregator,
+        resume_from_backup=False,
+    )
+    with db_v5.conn.read_ctx() as cursor:
+        cursor.execute("SELECT value FROM settings where name='active_modules'")
+        old_active_modules = json.loads(cursor.fetchone()[0])
```
Comment (LefterisJP) [inline rotkehlchen/tests/db/test_db_upgrades.py:2611]: please do an assertion that balancer is in here as a module

File: rotkehlchen/tests/unit/decoders/test_balancer.py [hunk 13]
```diff
@@ -639,3 +640,156 @@ def test_balancer_v1_join_gnosis(gnosis_inquirer, gnosis_accounts):
...
+    expected_events = [
+        EvmEvent(
+            tx_hash=tx_hash,
+            sequence_index=0,
+            timestamp=timestamp,
+            location=Location.ETHEREUM,
+            event_type=HistoryEventType.SPEND,
+            event_subtype=HistoryEventSubType.FEE,
+            asset=A_ETH,
+            balance=Balance(amount=FVal(gas_str)),
```
Comment (LefterisJP) [inline rotkehlchen/tests/unit/decoders/test_balancer.py:660]: For single args no need to do kwargs. Here and elsewhere

```suggestion
            balance=Balance(FVal(gas_str)),
```

File: rotkehlchen/db/upgrades/v45_v46.py (hunk 11 above)
Comment (prettyirrelevant) [reply rotkehlchen/db/upgrades/v45_v46.py:36]: yeah, it's just **balancer**. the ethereum module didn't have that difference.

File: rotkehlchen/chain/evm/decoding/balancer/mixins.py (hunk 2 above)
Comment (prettyirrelevant) [reply rotkehlchen/chain/evm/decoding/balancer/mixins.py]: fixed it.

**Review by LefterisJP — APPROVED** [review]
LGTM

**Pull request merged**